crypto/x509/x509_vfy.c

```diff
@@ -314,7 +322,7 @@ static X509 *find_issuer(X509_STORE_CTX *ctx, STACK_OF(X509) *sk, X509 *x)
 
     for (i = 0; i < sk_X509_num(sk); i++) {
         issuer = sk_X509_value(sk, i);
-        if (ctx->check_issued(ctx, x, issuer)) {
+        if (issuer != x && ctx->check_issued(ctx, x, issuer)) {
```

Comment:
why do we need this?
and if we need this, why not also add X509_cmp here?

Reply:
I've put this specific change into a commit of its own: 1358285dcc2472d6bec8bfa3098956af6bd6c2f1
with the following message:
```
    optimization and safety precaution in find_issuer() of x509_vfy.c:
    candidate issuer cert cannot be the same as the subject cert 'x'
```

So it is not strictly needed but is (at least) an optimization because when building up a chain, which is the context where this private function is used. In this context, it would be wrong to add an element to the chain where `issuer` is identical to `x` (the current cert). Likely the call to `check_issued(ctx, x, issuer)` will fail anyway in case `x == issuer`. If so, it is cheaper to do the equality check beforehand, and if not, this may incur the danger of including loops in the chain (as far as other measures to prevent this might not be sufficient in certain situations). 

As far as I understand, the more complex `X509_cmp()` is not needed in this context because of each cert there is likely only one instance/copy. And even if this does not hold, there is still the call to `check_issued(ctx, x, issuer)` to take care of that.

@vdukhovni, since you are the expert here, may I ask for your comment as well?

Reply:
I don't see duplicate suppression here doing any harm.  But Bernd is right that if we really wanted duplicate elimination, we'd have do `X509_cmp()`.  So this is doing a quick-n-dirty variant.

Mind you this function is ultimately called by `get_issuer`, where in 1.1.1 at least the comment above the function says `without duplicate elimination`, but I don't see anything in the code that depends on that property. So this change looks harmless to me. Please double-check that I did not miss anything...

Reply:
What about `get_issuer`?